src/WebJobs.Script.WebHost/Management/LegionInstanceManager.cs

```diff
@@ -0,0 +1,146 @@
+            IEnvironment environment, ILogger<LegionInstanceManager> logger, IMetricsLogger metricsLogger, IMeshServiceClient meshServiceClient)
+        {
+            _client = httpClientFactory?.CreateClient() ?? throw new ArgumentNullException(nameof(httpClientFactory));
+            _webHostEnvironment = webHostEnvironment ?? throw new ArgumentNullException(nameof(webHostEnvironment));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _metricsLogger = metricsLogger ?? throw new ArgumentNullException(nameof(metricsLogger));
+            _meshServiceClient = meshServiceClient ?? throw new ArgumentNullException(nameof(meshServiceClient));
+            _environment = environment ?? throw new ArgumentNullException(nameof(environment));
+        }
+
+        public Task<string> SpecializeMSISidecar(HostAssignmentContext context)
+        {
+            // Legion will take care of MSI Specialization
+            return null;
+        }
+
+        public bool StartAssignment(HostAssignmentContext context)
+        {
+            if (!_webHostEnvironment.InStandbyMode && !context.Environment.TryGetValue(EnvironmentSettingNames.ContainerStartContext, out string startContext))
+            {
+                // This is only true when specializing pinned containers.
+                _logger.LogError("Assign called while host is not in placeholder mode and start context is not present.");
+                return false;
+            }
+
+            if (_environment.IsContainerReady())
+            {
+                _logger.LogError("Assign called while container is marked as specialized.");
+                return false;
+            }
+
+            if (context.IsWarmupRequest)
+            {
+                // No need to download since file already exists
+                return true;
+            }
+
+            if (_assignmentContext == null)
+            {
+                lock (_assignmentLock)
+                {
+                    if (_assignmentContext != null)
+                    {
+                        return _assignmentContext.Equals(context);
+                    }
+                    _assignmentContext = context;
+                }
+
+                _logger.LogInformation($"Starting Assignment. Cloud Name: {_environment.GetCloudName()}");
+
+                // set a flag which will cause any incoming http requests to buffer
+                // until specialization is complete
+                // the host is guaranteed not to receive any requests until AFTER assign
+                // has been initiated, so setting this flag here is sufficient to ensure
+                // that any subsequent incoming requests while the assign is in progress
+                // will be delayed until complete
+                _webHostEnvironment.DelayRequests();
+
+                // start the specialization process in the background
+                Task.Run(async () => await Assign(context));
```

Comment:
Small nit: don't think you need `async`/`await` here.
```suggestion
                Task.Run() => Assign(context));
```

Reply:
But the Assign() method invokes the async _meshServiceClient.NotifyHealthEvent() task. Don't we need to await this task?

Reply:
It will still be awaited. `Task.Run` just wants a Task-returning method passed in. `() => Assign(context)` and `async () => await Assign(context)` have the same `Task` returning signature. Async-await is a compiler directive to generate a state machine to make task continuations easier. But in this case you have no continuation (specifically, your lambda does not do any work after `Assign` is finished). So you are safe to just directly pass the returned `Task` from `Assign` up to `Task.Run` and let it await it.